### ISSUE TITLE
Make MKL optional

### ISSUE THREAD

**Issue by jakirkham** [ContinuumIO/anaconda-issues#672]
Cont'd from this issue ( https://github.com/conda/conda/issues/2097 ).

It seems MKL wants to always install itself by default. This causes problems when building recipes that try to link against another BLAS (e.g. OpenBLAS). So, these now have to be designed to somehow keep MKL out, as well. This unfortunately is not simple and often requires ugly hacks that break in different ways. Instead it seem it would be simpler to just prefer the `nomkl` variant of packages by default unless `mkl` is explicitly requested.

**Comment by arogozhnikov** [comment]
> just prefer the nomkl variant of packages by default unless mkl is explicitly requested.
+1, now I am adding `nomkl` everywhere to avoid possible problems with linking (and licensing) - probably not what was expected. 

I am sure it was times just to propose adding `mkl` flag for those who are interested in speed.

**Comment by jakirkham** [comment]
Yeah, this comes at a difficult time as features (how `mkl`/`nomkl` operate) are being reworked. Though, on a positive note, there have been a lot of people hard at work trying to address these issues. There are already some significant improvements to the solver and it looks like a version 4 beta of `conda` is planned to come out soon that will include these changes and possibly a few more.

**Comment by jseabold** [comment]
+1 having just added nomkl to every single one of my build recipes to try really hard to avoid the 300mb download on each package build.

**Comment by gdementen** [comment]
I am bit late to the party, but this has just bitten me pretty hard. The problem is that I distribute executables built using a conda-installed python/numpy (using cx_freeze or pyinstaller), and I do not have a license to redistribute the MKL. So it would be great if nomkl worked on Windows too.

By the way, in issue conda/conda#2097, @ilanschnell told:

> On Windows, we have always been linking against MKL. However, with the Anaconda 2.5 release we separated the MKL runtime into its own conda package, in order to things uniformly on all platforms.

... but this is not true for the free packages...

As far as I can tell, the only versions of the package which bundle/require the MKL to run are those for 
Python3.5 (starting at numpy 1.9.2).
For all other versions of Python, I do not think any free version ever bundled/linked against the MKL, (except starting from 1.10.4, where all packages link against the external mkl package).

If you look at: http://repo.continuum.io/pkgs/free/win-64/ you immediately see which is which by looking at the file sizes. The 7 to 25Mb versions are those not linked with MKL, the 99Mb is packages bundle the MKL and the 2 to 3 Mb packages link against the external MKL package.

**Comment by gdementen** [comment]
From reading, https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/591215, it seems like I can get a community license which would allow me to redistribute executables for free (just by registering). The problem is still that my programs, when indirectly using the MKL are ridiculously large for no added benefit to me or my users (for some reason, my numexpr heavy code is not any faster when using the MKL).
